web/modules/custom/rate_limit/src/EventSubscriber/RateLimitSubscriber.php

```diff
@@ -0,0 +1,172 @@
+
+/**
+ * Event subscriber to limit api request in a time limit.
+ */
+final class RateLimitSubscriber implements EventSubscriberInterface {
+  /**
+   * The database service.
+   *
+   * @var \Drupal\Core\Database\Connection
+   */
+  protected $connection;
+  /**
+   * The config factory service.
+   *
+   * @var \Drupal\Core\Config\ConfigFactory
+   */
+  protected $config;
+
+  /**
+   * Constructs a RateLimitSubscriber object.
+   */
+  public function __construct(Connection $connection, ConfigFactory $config) {
+    $this->connection = $connection;
+    $this->config = $config;
+  }
+
+  /**
+   * Kernel request event handler.
+   *
+   * @throws \Symfony\Component\HttpKernel\Exception\TooManyRequestsHttpException
+   *   In case the rate limit exceeds.
+   */
+  public function onKernelRequest(RequestEvent $event): void {
+    $request = $event->getRequest();
+    $requestFormat = $event->getRequest()->getRequestFormat();
+    $clientIp = $request->getClientIp();
+    $config = $this->config->get(SettingsForm::CONFIG);
+    $rateLimit = $config->get("rate_limit");
+    $timeLimit = $config->get("time_limit");
+    if ($requestFormat !== "json" || !$rateLimit || !$timeLimit) {
+      return;
+    }
+    $didIpCrossRateLimit = $this->checkIfIpCrossedRateLimit($clientIp, (int) $rateLimit, (int) $timeLimit);
+    if ($didIpCrossRateLimit !== FALSE) {
+      throw new TooManyRequestsHttpException(0, "Rate limit exceeded. Please wait for another $didIpCrossRateLimit seconds before making more requests.");
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function getSubscribedEvents(): array {
+    return [
+      KernelEvents::REQUEST => ['onKernelRequest', 30],
+    ];
+  }
+
+  /**
+   * The function to check if the current IP has reached the rate limit.
+   *
+   * @param string $ip
+   *   The client ip of the request.
+   * @param int $rateLimit
+   *   The number of times an IP is allowed to make
+   *   the api call before it rejects.
+   * @param int $timeLimit
+   *   The time limit within which the rate limit applies.
+   *
+   * @return false|int
+   *   Returns false if the IP has not reached the limit yet,
+   *   else returns the time until it resets.
+   */
+  protected function checkIfIpCrossedRateLimit(string $ip, int $rateLimit = 5, int $timeLimit = 30) {
+
+    $ipDetails = $this->getIpDetails($ip);
+    if (empty($ipDetails)) {
+      $this->createNewIp($ip);
+    }
+    else {
+      ['timestamp' => $timestamp, 'count' => $count] = (array) array_values($ipDetails)[0];
+      $currentTimestamp = time();
+      if ($currentTimestamp < $timestamp + $timeLimit) {
+        $count++;
+        if ($count > $rateLimit) {
+          return $timestamp + $timeLimit - $currentTimestamp;
+        }
+        else {
+          $this->updateIpDetails($ip, ['count' => $count]);
+        }
+      }
+      else {
+        $this->updateIpDetails($ip, ['timestamp' => time(), 'count' => 1]);
+      }
+    }
```

Comment:
@SohamAxel inline comments always help us with code-readability. Please take care of it with your new solutions.